### PR TITLE
Provide Static Metric Macros

### PR DESCRIPTION
**Pull request by breeswish**
Moved from https://github.com/pingcap/rust-prometheus/pull/159 . Due to GitHub's unknown bug, previous PR's diff is buggy. Opening this new PR so that everything keeps fine.

Ref: buggy diff

image

### CODE REVIEW

**Comment by breeswish** [comment]
@siddontang Benchmark added.

```
test bench_counter_vec                      ... bench:          59 ns/iter (+/- 9)
test bench_single_counter                   ... bench:          11 ns/iter (+/- 1)
test bench_static_metrics_handwrite_1       ... bench:          12 ns/iter (+/- 1)
test bench_static_metrics_handwrite_2       ... bench:          12 ns/iter (+/- 1)
test bench_static_metrics_macro             ... bench:          12 ns/iter (+/- 1)
test bench_static_metrics_macro_with_lookup ... bench:          15 ns/iter (+/- 5)
```

As you can see, static metric macro achieves the same performance as hand-written static metrics, which is also as fast as native `Counter`s. Even dynamically looked up static metrics are (nearly) as fast as native `Counter`s. All of them beats `CounterVec`.

**Comment by breeswish** [comment]
@Hoverbear 

To simplify your review, I post some of our static metric design documentation which was discussed internally here.

The basic idea is to provide a macro which transforms the following syntax:

```rust
make_static_metric! {
    pub struct ExampleStaticCounter: Counter {
        "method" => {
            post,
            get,
            put,
            delete,
        },
        "test" => {
            foo,
            bar,
        },
        "server_ip" => {
            loop_back: "127.0.0.1",
            my_ip: MY_IP_ADDRESS,
            proxy_ip,
        },
    }
}

lazy_static! {
    pub static ref EXAMPLE_COUNTER_VEC: CounterVec =
        register_counter_vec!(
            "http_requests",
            "Total number of HTTP requests.",
            &["method", "test", "server_ip"]
        ).unwrap();

    pub static ref EXAMPLE_STATIC_COUNTER_VEC: ExampleStaticCounter = ExampleStaticCounter::from(&EXAMPLE_COUNTER_VEC);
}
```

Into the following Rust code:

```rust
pub use self::peometheus_scoped_foo::ExampleStaticCounter;

mod peometheus_scoped_foo {
    use super::*;
    use prometheus::Counter;
    use prometheus::CounterVec;
    use std::collections::HashMap;

    pub struct ExampleStaticCounter {
        pub post: ExampleStaticCounter2,
        pub get: ExampleStaticCounter2,
        pub put: ExampleStaticCounter2,
        pub delete: ExampleStaticCounter2,
    }

    impl ExampleStaticCounter {
        pub fn from(m: &CounterVec) -> ExampleStaticCounter {
            ExampleStaticCounter {
                post: ExampleStaticCounter2::from("post", m),
                get: ExampleStaticCounter2::from("get", m),
                put: ExampleStaticCounter2::from("put", m),
                delete: ExampleStaticCounter2::from("delete", m),
            }
        }

        pub fn get_by_label(&self, label: &str) -> &ExampleStaticCounter2 {
            match label {
                "post" => &self.post,
                "get" => &self.get,
                "put" => &self.put,
                "delete" => &self.delete,
                _ => panic!("unknown field"),
            }
        }
    }

    pub struct ExampleStaticCounter2 {
        pub foo: ExampleStaticCounter3,
        pub bar: ExampleStaticCounter3,
    }

    impl ExampleStaticCounter2 {
        pub fn from(label_1: &str, m: &CounterVec) -> ExampleStaticCounter2 {
            ExampleStaticCounter2 {
                foo: ExampleStaticCounter3::from(label_1, "foo", m),
                bar: ExampleStaticCounter3::from(label_1, "bar", m),
            }
        }

        pub fn get_by_label(&self, label: &str) -> &ExampleStaticCounter3 {
            match label {
                "foo" => &self.foo,
                "bar" => &self.bar,
                _ => panic!("unknown field"),
            }
        }
    }

    pub struct ExampleStaticCounter3 {
        pub loop_back: Counter,
        pub my_ip: Counter,
        pub proxy_ip: Counter,
    }

    impl ExampleStaticCounter3 {
        fn from(label_1: &str, label_2: &str, m: &CounterVec) -> ExampleStaticCounter3 {
            ExampleStaticCounter3 {
                loop_back: m.with(&{
                    let mut collection = HashMap::new();
                    collection.insert("method", label_1);
                    collection.insert("test", label_2);
                    collection.insert("server_ip", "127.0.0.1");
                    collection
                }),
                my_ip: m.with(&{
                    let mut collection = HashMap::new();
                    collection.insert("method", label_1);
                    collection.insert("test", label_2);
                    collection.insert("server_ip", MY_IP_ADDRESS);
                    collection
                }),
                proxy_ip: m.with(&{
                    let mut collection = HashMap::new();
                    collection.insert("method", label_1);
                    collection.insert("test", label_2);
                    collection.insert("server_ip", "proxy_ip");
                    collection
                }),
            }
        }

        pub fn get_by_label(&self, label: &str) -> &Counter {
            match label {
                "loop_back" => &self.loop_back,
                "my_ip" => &self.my_ip,
                "proxy_ip" => &self.proxy_ip,
                _ => panic!("unknown field"),
            }
        }
    }
}
```

**Comment by siddontang** [comment]
I guess Rust does some optimization for the deeply nested structs, maybe make them flat? Or pre know the field offset?

**Comment by breeswish** [comment]
@siddontang Yes, the struct is flat since all fields are non-references and the field offset is known in compile time, so that nesting is optimized out:

```rust
    unsafe {
        breakpoint();
    }

    let val10 = static_counter
        .val1
        .val2
        .val3
        .val4
        .val5
        .val6
        .val7
        .val8
        .val9
        .val10;

    unsafe {
        breakpoint();
    }
```

results in

```
    0x1000028db <+555>:  int3   
->  0x1000028dc <+556>:  mov    qword ptr [rbp - 0x78], rcx
    0x1000028e0 <+560>:  int3   
```

**Comment by breeswish** [comment]
@overvenus @siddontang  Updates?

**Comment by overvenus** [comment]
Sorry for the late reply! I will review it later this week, I promise!

File: static-metric/src/builder.rs [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+        let member_type = &self.member_type;
+        let values_str: Vec<_> = self.label.values.iter().map(|v| &v.value).collect();
+        let names_ident: Vec<_> = self.label.values.iter().map(|v| &v.name).collect();
+        quote!{
+            pub fn get(&self, value: &str) -> &#member_type {
+                match value {
+                    #(
+                        #values_str => &self.#names_ident,
+                    )*
+                    _ => panic!("unknown field `{}`", value),
```
Comment (overvenus) [inline static-metric/src/builder.rs]: So caution should be taken when we `get`.

File: static-metric/README.md [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    let vec = CounterVec::new(Opts::new("foo", "bar"), &["method", "product"]).unwrap();
+    let static_counter_vec = MyStaticCounterVec::from(&vec);
+
+    static_counter_vec.post.foo.inc();
+    static_counter_vec.delete.bar.inc_by(4.0);
+    assert_eq!(static_counter_vec.post.bar.get(), 0.0);
+    assert_eq!(vec.with_label_values(&["post", "foo"]).get(), 1.0);
+    assert_eq!(vec.with_label_values(&["delete", "bar"]).get(), 4.0);
+}
+```
```
Comment (overvenus) [inline static-metric/README.md]: Maybe we need a `CAVEATS` section about panic.

File: static-metric/src/builder.rs (hunk 1 above)
Comment (Hoverbear) [reply static-metric/src/builder.rs]: Perhaps it should return an `Option<T>` instead?

Comment (breeswish) [reply static-metric/src/builder.rs]: hmmmm.. Good idea. We can change to `Option` to leave the choice of panic to users.

**Comment by breeswish** [comment]
@overvenus Updated to use `Option<_>`, please kindly give your approves :)

**Pull request merged**